server-spi/src/main/java/org/keycloak/models/KeycloakSession.java

```diff
@@ -243,4 +245,6 @@ default <T> T getAttributeOrDefault(String attribute, T defaultValue) {
      */
     ClientPolicyManager clientPolicy();
 
+    boolean isClosed();
```

Comment:
If calling the close method twice will produce an exception, I'd say exposing this is much friendlier than making callers use try catch blocks

Somewhat related to this, some of the multi-thread concerns of the KeycloakSession seem unclear. There's synchronization in several places, but it's not complete, nor is it used on the close method.

Reply:
Sorry, I should be specifying KeycloakContext instead - but the same thinking should apply since it's 1-1 with the session. https://github.com/keycloak/keycloak/blob/be32f8b1bfb4fa709aa48552a1bbfdcf9fab6299/services/src/main/java/org/keycloak/services/DefaultKeycloakContext.java#L123 - there's synchronization on several get methods

Reply:
I would say that the synchronization block does not make sense there. AFAIK, the client connection is set once per-thread and, as you said, bound with a session for its entire lifecycle.

Reply:
Ok, I'll remove them from this proposal, and we can separate that out to a different pr if this one does not move forward.